Prometheus pushgateway support

lwsanty opened the pull request:
1) done refactoring: storage client factory, storage client interface with implementations
2) implemented prometheus pushgateway client
closes #4

Signed-off-by: lwsanty <lwsanty@gmail.com>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bblfsh/performance/6)
<!-- Reviewable:end -->

kuba-- commented on cmd/bblfsh-performance/endtoend/endtoend.go:
```diff
@@ -12,6 +13,9 @@ import (
 	"github.com/bblfsh/performance/util"
 	"github.com/bblfsh/performance/util/docker"
 	"github.com/bblfsh/performance/util/storage"
+	"github.com/bblfsh/performance/util/storage/influxdb"
```
IMO we don't need `util` sub-package here. I suggest to put it directly in `performance`

lwsanty replied:
Done

kuba-- commented on README.md:
```diff
@@ -5,29 +5,36 @@
 export INFLUX_DB=mydb
 export INFLUX_MEASUREMENT=benchmark
-bblfsh-performance parse-and-store --language=go --commit=3d9682b /var/log/bench0 /var/log/bench1
+bblfsh-performance parse-and-store --language=go --commit=3d9682b --storage="influxdb" /var/log/bench0 /var/log/bench1
```
If you don't specify commit, I think we can take the `HEAD`

lwsanty replied:
I think your comment somehow relates to this thread https://reviewable.io/reviews/bblfsh/performance/1#-LjKkfkM8v5R3iKGtMKR
`bblfsh-performance` does not know/interact with git(yet) and primary aim is to run it in CI, anyways it can be done like `--commit=$(git rev-parse HEAD)`

Or are you talking about `HEAD` as a string?..

kuba-- replied:
just thought. You can also use `go-git` to resolve `HEAD` to concrete commit hash.

kuba-- replied:
but generally it's not super important (must have).

lwsanty replied:
filed issue on that https://github.com/bblfsh/performance/issues/7

kuba-- commented on util/common.go:
```diff
@@ -57,6 +58,17 @@ func BenchmarkResultToParseBenchmark(name string, b *testing.BenchmarkResult) *p
+// Example: BenchmarkGoDriver/transform/accumulator_factory-4 -> accumulator_factory
+func ParseBenchmarkName(name string) string {
+	spl := strings.Split(name, "/")
+	name = spl[len(spl)-1]
```
`filepath.Base("BenchmarkGoDriver/transform/accumulator_factory-4")`

lwsanty replied:
initially it was done using `filepath.Base` but
https://reviewable.io/reviews/bblfsh/performance/1#-Lj1YiAXCy6I0H4XMjGi

kuba-- replied:
@dennwc  - why `filepath.Base` is incorrect here? We have a absolute path here and we wanna get a benchmark file name. If we rename name param to `path` and function name to `benchamarkName` I think it will be - WDYT?

kuba-- commented on util/docker/docker.go:
```diff
@@ -23,7 +23,7 @@ var (
 	errConnectToDockerFailed = errors.NewKind("could not connect to docker")
 	errResourceStartFailed   = errors.NewKind("could not start resource")
-	errPortWaitTimeout       = errors.NewKind("could not wait until port %d is enabled")
+	errPortWaitTimeout       = errors.NewKind("could not wait until port %s is enabled")
```
just in case I suggest to use `%v`, so you can pass string or number here and it will work.

lwsanty replied:
Done

kuba-- commented on util/storage/prom-pushgateway/pushgateway.go:
```diff
@@ -0,0 +1,82 @@
+	for _, b := range benchmarks {
+		tmpValues := append([]string{util.ParseBenchmarkName(b.Name)}, values...)
+
+		metrics[storage.PerOpSeconds].WithLabelValues(tmpValues...).Observe(float64(b.NsPerOp / 1e9))
```
here I suggest to use `time.Duration(b.NsPerOp).Seconds()`

lwsanty replied:
Done

kuba-- commented on util/common.go:
```diff
@@ -67,3 +79,16 @@ func WrapErr(err error, kinds ...*errors.Kind) error {
+		keys = append(keys, k)
+		values = append(values, v)
+	}
+	return keys, values
```
Just I want to be sure that it's OK that order of returned slices can be random. in other words sometimes `SplitStringMap(mymap) != SplitStringMap(mymap)`

lwsanty replied:
I think it's ok because:
- there're tags and order is not crucial
- keys and values order are preserved

bzz commented on go.sum:
```diff
@@ -89,11 +107,24 @@ github.com/pkg/errors v0.8.0/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINE
 github.com/sirupsen/logrus v1.0.5 h1:8c8b5uO0zS4X6RPl/sd1ENwSkIc0/H2PaHxE3udaE8I=
 github.com/sirupsen/logrus v1.0.5/go.mod h1:pMByvHTf9Beacp5x1UXfOR9xyW/9antXMhjMPG0dEzc=
+github.com/sirupsen/logrus v1.2.0 h1:juTguoYk5qI21pwyTXY3B3Y5cOTH3ZUyZCg1v/mihuo=
+github.com/sirupsen/logrus v1.2.0/go.mod h1:LxeOpSwHxABJmUn/MG1IvRgCAasNZTLOkJPxbbu5VWo=
```
Wondering, if that is after `go mod tidy`?

lwsanty replied:
probably goland ide executes `go mod tidy`

bzz commented on util/storage/influxdb/influxdb.go:
```diff
@@ -0,0 +1,121 @@
+
+	"github.com/bblfsh/performance/util"
+	"github.com/bblfsh/performance/util/storage"
+	"github.com/orourkedd/influxdb1-client/client"
```
A nit pick here and below - we usually try to visually separate groups of imports: stdlib, same repository, all the rest.

lwsanty replied:
on previous projects we had a rule 1) std 2) all the others :)
Done.

bzz commented on util/storage/storage.go:
```diff
@@ -1,121 +1,60 @@
-	errDumpFailed      = errors.NewKind("cannot dump batch points")
-)
+	// Constructors is a map of all supported storage client constructors
+	Constructors = make(map[string]Constructor)
```
If it is used only inside the single package, I wonder if it has to be exported here?

lwsanty replied:
Done

bzz reviewed (approved):
Looks great to me, as soon as few suggestions above are addressed

dennwc reviewed (changes requested):
> Reviewed 6 of 12 files at r1, 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/bblfsh/performance/6)* status: all files reviewed, 14 unresolved discussions (waiting on @bzz, @creachadair, @kuba--, and @lwsanty)

___
*[storage/influxdb/influxdb.go, line 80 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVO8EC11uzFRbX8HdN:-LjVO8EC11uzFRbX8HdO:b-ypezgs) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/storage/influxdb/influxdb.go#L80)):*
> ```Go
> 	eventTime := time.Now()
> 	for _, b := range benchmarks {
> 		tags["name"] = performance.ParseBenchmarkName(b.Name)
> ```

Since it will be called by each an every `storage`, maybe we should create a constructor for `parse.Benchmark` that does this.
___
*[storage/influxdb/influxdb.go, line 112 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVOSn6437kdnpcejDW:-LjVOSn6437kdnpcejDX:b-lhh5mg) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/storage/influxdb/influxdb.go#L112)):*
> ```Go
> }
> 
> func influxDBClient(conf influxConfig) (client.Client, error) {
> ```

Inline into `NewClient`?
___
*[storage/std/std.go, line 13 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVOpfxERBkURVJAJb2:-LjVOpfxERBkURVJAJb3:bgzncvi) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/storage/std/std.go#L13)):*
> ```Go
> 
> // Kind is a string that represents stdout
> const Kind = "std"
> ```

Call it `file` maybe? Stdout is a file as well, and I don't see any reason this cannot dump to an arbitrary file on the disk (in the future).
___
*[util/storage/influxdb/influxdb.go, line 1 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVPO1Q9fcteeNiYkcD:-LjVPO1Q9fcteeNiYkcE:bd5bpx9) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/util/storage/influxdb/influxdb.go#L1)):*
> ```Go
> package influxdb
> ```

How it's different from `storage/influxdb`? Looks like a duplicate file.
___
*[storage/storage.go, line 53 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVNGQ7-6FaumsdMYBb:-LjVNGQ82rVV7Al6peSE:bhngb99) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/storage/storage.go#L53)):*
> ```Go
> // so kind can be checked much earlier then storage client acquired
> // and prevent the situation when tests passed and store failed because kind is not supported
> func ValidateKind(kind string) (constructor, error) {
> ```

Either unexport the function or export the `contructor`.
___
*[util/common.go, line 65 at r1](https://reviewable.io/reviews/bblfsh/performance/6#-LjVKmEjTcW7P_P4-1BR-r1-65:-LjVKrhT54UJsIjV0SZL:bb8dt8x) ([raw file](https://github.com/bblfsh/performance/blob/fd7da60fb2ceb51f2815519bcbf7b61fce59152a/util/common.go#L65)):*
<details><summary><i>Previously, kuba-- (Kuba Podgórski) wrote…</i></summary><blockquote>

@dennwc  - why `filepath.Base` is incorrect here? We have a absolute path here and we wanna get a benchmark file name. If we rename name param to `path` and function name to `benchamarkName` I think it will be - WDYT?
</blockquote></details>

The thing is, those are not paths. For example, paths on Windows should have a backslash (`\`), and `filepath` will try to split on `\` instead. Go test names on the other hand always use `/`. Of course, we don't expect this tool to work on Windows, but it still makes the code semantically incorrect.

It is okay to use `path.Base()` though.
___
*[util/common.go, line 93 at r1](https://reviewable.io/reviews/bblfsh/performance/6#-LjVKmEjTcW7P_P4-1BR-r1-93:-LjVNivU7fnjOU22A6F4:b-wut429) ([raw file](https://github.com/bblfsh/performance/blob/fd7da60fb2ceb51f2815519bcbf7b61fce59152a/util/common.go#L93)):*
<details><summary><i>Previously, lwsanty wrote…</i></summary><blockquote>

I think it's ok because:
- there're tags and order is not crucial
- keys and values order are preserved
</blockquote></details>

I also think it makes sense to sort it. The code that uses it passes it as a vector to find a corresponding metric, that in terms uses it as a "hash" for a lookup. It may or may not sort them, so it's better to make sure that we do it on our side.


<!-- Sent from Reviewable.io -->

lwsanty reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/bblfsh/performance/6)* status: all files reviewed, 14 unresolved discussions (waiting on @bzz, @creachadair, @dennwc, and @kuba--)

___
*[storage/influxdb/influxdb.go, line 80 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVO8EC11uzFRbX8HdN:-LjVwk3P7NBMGJVx1lXV:b-896fix) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/storage/influxdb/influxdb.go#L80)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Since it will be called by each an every `storage`, maybe we should create a constructor for `parse.Benchmark` that does this.
</blockquote></details>

Done.
___
*[storage/influxdb/influxdb.go, line 112 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVOSn6437kdnpcejDW:-LjVr1v94yChSWEZ6Bl5:b-896fix) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/storage/influxdb/influxdb.go#L112)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Inline into `NewClient`?
</blockquote></details>

Done.
___
*[storage/std/std.go, line 13 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVOpfxERBkURVJAJb2:-LjVrx4T0F25LL6XwfDV:b-896fix) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/storage/std/std.go#L13)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Call it `file` maybe? Stdout is a file as well, and I don't see any reason this cannot dump to an arbitrary file on the disk (in the future).
</blockquote></details>

Done.
___
*[util/storage/influxdb/influxdb.go, line 1 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVPO1Q9fcteeNiYkcD:-LjVcWS87U9eCjRHgPaR:bro3itv) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/util/storage/influxdb/influxdb.go#L1)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

How it's different from `storage/influxdb`? Looks like a duplicate file.
</blockquote></details>

Done. Some problems with my git, each commit it says that it has merge conflict with previous one
___
*[storage/storage.go, line 53 at r2](https://reviewable.io/reviews/bblfsh/performance/6#-LjVNGQ7-6FaumsdMYBb:-LjVWvL93ehFFB1AycBK:b-896fix) ([raw file](https://github.com/bblfsh/performance/blob/08a5cf07517320b9e621cc302395330489cdd288/storage/storage.go#L53)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Either unexport the function or export the `contructor`.
</blockquote></details>

Done.
___
*[util/common.go, line 65 at r1](https://reviewable.io/reviews/bblfsh/performance/6#-LjVKmEjTcW7P_P4-1BR-r1-65:-LjVWFHQDdoE2BGMsFt5:b1c84nb) ([raw file](https://github.com/bblfsh/performance/blob/fd7da60fb2ceb51f2815519bcbf7b61fce59152a/util/common.go#L65)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

The thing is, those are not paths. For example, paths on Windows should have a backslash (`\`), and `filepath` will try to split on `\` instead. Go test names on the other hand always use `/`. Of course, we don't expect this tool to work on Windows, but it still makes the code semantically incorrect.

It is okay to use `path.Base()` though.
</blockquote></details>

So should I rewrite this part using `path.Base()` then?
___
*[util/common.go, line 93 at r1](https://reviewable.io/reviews/bblfsh/performance/6#-LjVKmEjTcW7P_P4-1BR-r1-93:-LjVq_HSA50NQosiWrMY:b-896fix) ([raw file](https://github.com/bblfsh/performance/blob/fd7da60fb2ceb51f2815519bcbf7b61fce59152a/util/common.go#L93)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

I also think it makes sense to sort it. The code that uses it passes it as a vector to find a corresponding metric, that in terms uses it as a "hash" for a lookup. It may or may not sort them, so it's better to make sure that we do it on our side.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dennwc reviewed (changes requested):
> Reviewed 12 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/bblfsh/performance/6)* status: all files reviewed, 10 unresolved discussions (waiting on @bzz, @creachadair, @kuba--, and @lwsanty)

___
*[common.go, line 105 at r3](https://reviewable.io/reviews/bblfsh/performance/6#-LjW3M_h8VchgVNexKAn:-LjW3M_h8VchgVNexKAo:bd63jn) ([raw file](https://github.com/bblfsh/performance/blob/1c1dbc9dc2f2be7b3d9cc400560d36188b111a59/common.go#L105)):*
> ```Go
> 	}
> 
> 	sort.Slice(keys, func(i, j int) bool {
> ```

`sort.Strings`
___
*[util/common.go, line 65 at r1](https://reviewable.io/reviews/bblfsh/performance/6#-LjVKmEjTcW7P_P4-1BR-r1-65:-LjW3ODk5s5AeyZhKnGj:bcnd7u6) ([raw file](https://github.com/bblfsh/performance/blob/fd7da60fb2ceb51f2815519bcbf7b61fce59152a/util/common.go#L65)):*
<details><summary><i>Previously, lwsanty wrote…</i></summary><blockquote>

So should I rewrite this part using `path.Base()` then?
</blockquote></details>

I'm okay with leaving it as-is.


<!-- Sent from Reviewable.io -->

lwsanty reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/bblfsh/performance/6)* status: all files reviewed, 10 unresolved discussions (waiting on @bzz, @creachadair, @dennwc, and @kuba--)

___
*[common.go, line 105 at r3](https://reviewable.io/reviews/bblfsh/performance/6#-LjW3M_h8VchgVNexKAn:-LjWa9Rw-2bBAGe_bshI:b-896fix) ([raw file](https://github.com/bblfsh/performance/blob/1c1dbc9dc2f2be7b3d9cc400560d36188b111a59/common.go#L105)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

`sort.Strings`
</blockquote></details>

Done.
___
*[util/common.go, line 65 at r1](https://reviewable.io/reviews/bblfsh/performance/6#-LjVKmEjTcW7P_P4-1BR-r1-65:-LjWaFUV6h66PVhc4HQv:b9dr70j) ([raw file](https://github.com/bblfsh/performance/blob/fd7da60fb2ceb51f2815519bcbf7b61fce59152a/util/common.go#L65)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

I'm okay with leaving it as-is.
</blockquote></details>

@kuba-- ?


<!-- Sent from Reviewable.io -->

dennwc reviewed (approved):
> Reviewed 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/bblfsh/performance/6)* status: all files reviewed, 9 unresolved discussions (waiting on @bzz, @creachadair, and @kuba--)



<!-- Sent from Reviewable.io -->

creachadair reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/bblfsh/performance/6)* status: all files reviewed, 11 unresolved discussions (waiting on @bzz, @kuba--, and @lwsanty)

___
*[common.go, line 28 at r4](https://reviewable.io/reviews/bblfsh/performance/6#-LjY9Cx_6V3Y-JHE7GK9:-LjY9A0KD23h4ygcQ9qB:b3uz8mc) ([raw file](https://github.com/bblfsh/performance/blob/4ee01868ae67d4abf14b08ca70404c024793f193/common.go#L28)):*
> ```Go
> // Benchmark is a wrapper around parse.Benchmark and serves for formatting and arranging data before storing
> type Benchmark struct {
> 	parse.Benchmark
> ```

Do we really want to re-export the same interface as the underlying `parse.Benchmark`? I'm trying to figure out where that would help—existing code that wants a `parse.Benchmark` doesn't know about the extension methods. 

It's not wrong to embed like this, but unless you want to use the new type as a value in an existing interface, I suggest defaulting to an exported but named field: `Benchmark *parse.Benchmark`. This still permits the underlying value to be dispatched into the existing libraries, but doesn't do it implicitly.

I didn't notice a case where we needed the embedded behaviour, but I might have missed something.
___
*[common.go, line 51 at r4](https://reviewable.io/reviews/bblfsh/performance/6#-LjYAPRB7Kom1dfXH8iK:-LjYAPRB7Kom1dfXH8iL:b3shlxn) ([raw file](https://github.com/bblfsh/performance/blob/4ee01868ae67d4abf14b08ca70404c024793f193/common.go#L51)):*
> ```Go
> // Example: BenchmarkGoDriver/transform/accumulator_factory-4 -> accumulator_factory
> func parseBenchmarkName(name string) string {
> 	spl := strings.Split(name, "/")
> ```

Is the goal here to pull off the last slash-delimited field? If so, I recommend:

```go
if i := strings.LastIndex(name, "/"); i >= 0 {
  name = name[i+1:]
}
if i := strings.LastIndexAny(name, "-."); i >= 0 {
  return name[:i]
}
return name
```

Rationale: We don't really need to split the whole string, we only need the first (or last) entry in each case.

But what you wrote here is also fine.


<!-- Sent from Reviewable.io -->

creachadair reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/bblfsh/performance/6)* status: all files reviewed, 11 unresolved discussions (waiting on @bzz, @kuba--, and @lwsanty)

___
*[common.go, line 51 at r4](https://reviewable.io/reviews/bblfsh/performance/6#-LjYAPRB7Kom1dfXH8iK:-LjYD5-i-LTvRohDD78O:bjyti3k) ([raw file](https://github.com/bblfsh/performance/blob/4ee01868ae67d4abf14b08ca70404c024793f193/common.go#L51)):*
<details><summary><i>Previously, creachadair (M. J. Fromberger) wrote…</i></summary><blockquote>

Is the goal here to pull off the last slash-delimited field? If so, I recommend:

```go
if i := strings.LastIndex(name, "/"); i >= 0 {
  name = name[i+1:]
}
if i := strings.LastIndexAny(name, "-."); i >= 0 {
  return name[:i]
}
return name
```

Rationale: We don't really need to split the whole string, we only need the first (or last) entry in each case.

But what you wrote here is also fine.
</blockquote></details>

Er, in the second case, I meant `strings.IndexAny`.


<!-- Sent from Reviewable.io -->

lwsanty committed:
prometheus pushgateway support
1) done refactoring: storage client factory, storage client interface with implementations
2) implemented prometheus pushgateway client
closes #4

Signed-off-by: lwsanty <lwsanty@gmail.com>
